clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/evaluation/EvaluationManager.kt

```diff
@@ -0,0 +1,503 @@
+     * @param inappNotifs A list of JSONObjects representing in-app notifications to be evaluated.
+     * @param clearResource A function to clear resources, invoked when a limit indicates data should be discarded.
+     *                     Accepts a URL parameter
+     *
+     * @return A list of JSONObjects representing in-app notifications that meet the criteria for display.
+     */
+    @VisibleForTesting
+    internal fun evaluate(
+        event: EventAdapter,
+        inappNotifs: List<JSONObject>,
+        clearResource: (url: String) -> Unit = {}
+    ): List<JSONObject> {
+        val eligibleInApps: MutableList<JSONObject> = mutableListOf()
+
+        for (inApp in inappNotifs) {
+            val campaignId = inApp.optString(Constants.INAPP_ID_IN_PAYLOAD)
+
+            val matchesTrigger =
+                triggersMatcher.matchEvent(getWhenTriggers(inApp), event)
+            if (matchesTrigger) {
+                Logger.v("INAPP", "Triggers matched for event ${event.eventName} against inApp $campaignId")
+                triggersManager.increment(campaignId)
+
+                val matchesLimits = limitsMatcher.matchWhenLimits(getWhenLimits(inApp), campaignId)
+                val discardData = limitsMatcher.shouldDiscard(getWhenLimits(inApp), campaignId)
+
+                if (discardData) {
+                    clearResource.invoke("") // todo pass correct url
+                }
+                if (matchesLimits) {
+                    Logger.v("INAPP", "Limits matched for event ${event.eventName} against inApp $campaignId")
+                    eligibleInApps.add(inApp)
+                } else {
+                    Logger.v("INAPP", "Limits did not matched for event ${event.eventName} against inApp $campaignId")
+                }
+            } else {
+                Logger.v("INAPP", "Triggers did not matched for event ${event.eventName} against inApp $campaignId")
+            }
+        }
+        return eligibleInApps
+    }
+
+    @VisibleForTesting
+    internal fun getWhenTriggers(triggerJson: JSONObject): List<TriggerAdapter> {
+        val whenTriggers = triggerJson.optJSONArray(Constants.INAPP_WHEN_TRIGGERS).orEmptyArray()
+        return (0 until whenTriggers.length()).mapNotNull {
+            val jsonObject = whenTriggers[it] as? JSONObject
+            jsonObject?.let { nonNullJsonObject -> TriggerAdapter(nonNullJsonObject) }
+        }
+    }
+
+    internal fun getWhenLimits(limitJSON: JSONObject): List<LimitAdapter> {
+        val frequencyLimits = limitJSON.optJSONArray(Constants.INAPP_FC_LIMITS).orEmptyArray()
+        val occurrenceLimits = limitJSON.optJSONArray(Constants.INAPP_OCCURRENCE_LIMITS).orEmptyArray()
+
+        return (frequencyLimits.toList<JSONObject>() + occurrenceLimits.toList()).mapNotNull {
+            if (it.isNotNullAndEmpty()) {
+                LimitAdapter(it)
+            } else null
+        }.toMutableList()
```

Comment:
.toMutableList() is not required

Reply:
@piyush-kukadiya

Reply:
added changes in https://github.com/CleverTap/clevertap-android-sdk/commit/c8cd859b1b2afb57d5b98fc9322c277d8a3bbaca